feat(junit): Implement automatic saving of traces and screenshots via fixtures

uchagani opened the pull request:
Resolves #1526.  This PR also adds saving screenshots.

I tried to keep the API of the fixture-generated artifacts as close to node as possible.  

Traces:
By default, if no `outputDir` is specified in the `Options`, the default output path will be `projectDir/test-results/fully.qualified.ClassName.testMethodName-browserName/trace.zip`.  If an `outputDir` is specified in the `Options` then the artifacts will go in `userSpecifiedOutputDir/fully.qualified.ClassName.testMethodName-browserName/trace.zip`.


Screenshots:
Screenshots follow the same nomenclature as the traces with the exception that screenshot files are named `test-finished-1.png`.  Depending on the number of pages a `BrowserContext` has, the `1` will be incremented per `Page`.

~~I added tests to test this feature.  Since traces and screenshots are created after the test is finished, I needed to add a new test dependency (`junit-platform-testkit`).   This is JUnit's preferred way to test extensions.~~

~~`@Transient` annotation was created because the new tests should not be run by themselves and only run through the `EngineTestKit`.  I updated the GitHub actions to exclude these tests.~~

~~There is some weird threading issues happening when the new tests run with the existing suite so I added a new step in the CI actions to run the new tests separately.~~

problems with running tests through the `junit-platform-testkit`: https://github.com/microsoft/playwright-java/pull/1560#issuecomment-2305981500

uchagani commented:
I don't think the failures here are related to the PR.  could we try re-running?

yury-s commented:
> I don't think the failures here are related to the PR. could we try re-running?

Restarted. But it looks like at least some of them are in the new tests.

yury-s commented on playwright/src/test/java/com/microsoft/playwright/junit/TestTracesAndScreenshots.java:
```diff
@@ -0,0 +1,201 @@
+
+    for (Method test : testMethods) {
+      File traceFile = getTracePathForDefaultLocation(testClass, test).toFile();
+      if (test.getName().contains("ShouldNot")) {
```
I'd check entire method name "traceShouldNotExistWhenTestPasses"  for clarity.

Also here and in other tests, when iterating over several methods, include the name in the message for easier debugging: assertFalse(traceFile.exists(), test.getName());

yury-s reviewed (approved):
Looks good, but let's figure out the multithreading issues and other failures of the new tests on the bots.

uchagani commented:
@yury-s the threading issues, I believe, are resulting from tests running via the engine kit and the normal way at the same time.

I believe the threading issues are due to use using threadlocal in our extensions. Somewhere, playwright is getting closed by the ExtensionContext but there are still tests running on that thread.

We should eventually move away from using threadlocal and store objects in the ExtensionContext (this is what Junit recommends). However, I'm not sure how we can scope to threads while doing that. We can scope to method or class easily but we want to reuse playwright and keep it open across test classes.


I don't think users will be affected by this but will verify by running the whole suite except engine kit tests.

uchagani commented on playwright/src/test/java/com/microsoft/playwright/junit/TestTracesAndScreenshots.java:
```diff
@@ -0,0 +1,201 @@
+
+// Test both traces and screenshots in the same class as they both have the same default directory
+// and cleanup is easier
+@Tag("fixtureTracing") // These tests should be run in isolation as there are threading issues when run with others
```
mentioned in the main comments, I think they have to do with tests being run in the EngineTestKit on the same thread as the rest of the tests.  I don't think our users will be affected by this but I will double check.

uchagani commented on playwright/src/test/java/com/microsoft/playwright/junit/TestTracesAndScreenshots.java:
```diff
@@ -0,0 +1,201 @@
+  }
+
+  public static Path getCustomOutputPath() {
+    return Paths.get(System.getProperty("java.io.tmpdir")).resolve("playwright-custom-output");
```
I tried using this but we need this resolved in the options which is before any test would resolve this, so i chose to do it this way and clean up manually.

uchagani commented:
@yury-s i believe i have addressed all the comments.  I also verified that the threading issues are not user-facing and should not affect users.  After this PR, I can start looking into how we can store playwright fixtures in the ExtensionContext and scope them to specific threads

uchagani commented:
will work on test failures this evening

uchagani commented:
I'm sorry for the delay life got in the way recently.  Hopefully will get to this this weekend

yury-s commented:
No problem at all, hope everything is alright.

uchagani commented:
@yury-s was wondering if it's okay to land this feature without tests.  The JUnit Engine Test Kit isn't playing well.  I have confirmed that the issue is not in the feature but how the engine test kit runs the tests.

yury-s commented on playwright/src/main/java/com/microsoft/playwright/impl/junit/BrowserContextExtension.java:
```diff
@@ -66,10 +64,135 @@ public static BrowserContext getOrCreateBrowserContext(ExtensionContext extensio
+  }
+
+  private void cleanupResources() {
+    closePage();
```
Can we rely on `BrowserContext.close()` closing all pages and avoid this call?

yury-s commented on playwright/src/main/java/com/microsoft/playwright/impl/junit/BrowserContextExtension.java:
```diff
@@ -66,10 +64,135 @@ public static BrowserContext getOrCreateBrowserContext(ExtensionContext extensio
+      Files.walk(outputPath).sorted(Comparator.reverseOrder()).map(Path::toFile).forEach(File::delete);
+      Files.createDirectories(outputPath);
+    } catch (NoSuchFileException ignored) {
+      // swallow
```
This doesn't seem right, maybe wrap deletion into try catch but move createDirectories outside?

uchagani replied:
I was doing too much.  I think it's cleaned up now where if the directory exists, I don't touch it.

yury-s commented:
> @yury-s was wondering if it's okay to land this feature without tests. The JUnit Engine Test Kit isn't playing well. I have confirmed that the issue is not in the feature but how the engine test kit runs the tests.

Yes, if it proves too difficult to test it, we can land it without tests. Would be nice to have some description of the problem here, so that later on if we wonder why it's difficult to write such tests we wouldn't need to repeat all the debugging and investigation that you've already done.

yury-s commented on playwright/src/main/java/com/microsoft/playwright/impl/junit/BrowserContextExtension.java:
```diff
@@ -66,10 +64,135 @@ public static BrowserContext getOrCreateBrowserContext(ExtensionContext extensio
+    }
+  }
+
+  private static void saveScreenshot(ExtensionContext extensionContext, boolean didTestFail) {
```
I'd remove it altogether and recommend tracing instead in all scenarios.

uchagani commented:
looking into failures

uchagani commented:
@yury-s All tests should be passing now
